include/zephyr/senss/senss.h

```diff
@@ -0,0 +1,384 @@
+	const char *name;
+
+	/** Friendly name of the sensor instance */
+	const char *friendly_name;
+
+	/** Vendor name of the sensor instance */
+	const char *vendor;
+
+	/** Model name of the sensor instance */
+	const char *model;
+
+	/** Sensor type */
+	int32_t type;
+
+	/** Sensor index for multiple sensors under same sensor type */
+	int32_t sensor_index;
+
+	/** Sensor flags */
+	uint32_t flags;
+
+	/** Minimal report interval in micro seconds  */
+	uint32_t minimal_interval;
+
+	/** Sensor version */
+	struct senss_sensor_version version;
+};
+
+
+
+/**
+ * @brief Initialize sensor subsystem runtime.
+ *
+ * This will allocate internal resources, such as create internal schedule
+ * task, create internal management data structures.
+ *
+ * It will also enumerate all registered sensor instances in sensor subsystem,
+ * build sensors report relational data model, if a virtual sensor's required dependency
+ * sensor type not found, will return error.
+ * Then call each sensor instances' init callback API.
+ *
+ * @return 0 on success or negative error value on failure.
+ *
+ */
+int senss_init(void);
+
+ /**
+  * @brief De-initialize sensor subsystem runtime.
+  *
+  * This will release all dynamically allocated internal resources, and call
+  * all sensor instances' deinit callback API.
+  *
+  * @return 0 on success or negative error value on failure.
+  */
+int senss_deinit(void);
+
+ /**
+  * @brief Get all supported sensor instances' information.
+  *
+  * This API just returned read only information of sensor instances,
+  * no side effect to sensor instances.
+  *
+  * @param **info For receiving sensor instances' information array pointer.
+  *
+  * @return Number of sensor instances, 0 returned if no sensors.
+  */
+int senss_get_sensors(const struct senss_sensor_info **info);
+
+
+/**
+ * @brief Open sensor instance by sensor type and sensor index.
+ *
+ * Application clients use it to open a sensor instance and get its handle.
+ * Support multiple Application clients for open same sensor instance,
+ * in this case, the returned handle will different for different clients.
+ *
+ * @param type The sensor type which need to open.
+ *
+ * @param sensor_index The index which need to open.
+ *
+ * @param *handle The opened instance handle, if failed will be set to -1.
+ *
+ * @return 0 on success or negative error value on failure.
+ */
+int senss_open_sensor(
+		int type, int sensor_index,
+		int *handle);
+
+/**
+ * @brief Close sensor instance.
+ *
+ * @param handle The sensor instance handle need to close.
+ *
+ * @return 0 on success or negative error value on failure.
+ */
+int senss_close_sensor(
+		int handle);
+
+
+/**
+ * @brief Get current report interval of given client handle.
+ *
+ * @param handle The sensor instance handle.
+ *
+ * @param value The data buffer to receive value (micro seconds unit).
+ *
+ * @return 0 on success or negative error value on failure.
+ */
+int senss_get_interval(
+		int handle,
```

Comment:
The interval should be arbitrated, the handle makes no difference, just the sensor index.

Reply:
sorry, as mentioned above, handle will be translated to connection, includes both reporter and client

Reply:
I understand that's what the `handle` is translated to, but I'm not seeing why it's needed. For example if I have 4 connections open to the sensor at lets say 1, 2, 4, and 8 kHz ODRs, getting the interval for that sensor should just return 8kHz since that's the rate it's actually set to. The `handle` makes no difference for getters.

Reply:
this API returns the per client config report interval value (return default value if not configured), not the value after arbitration, so the `handle` is needed. 

When client A sets 10 Hz report interval to a sensor, even there is another client B sets 100 Hz to this same sensor, and internally the sensor works as 100 Hz after arbitration, from client A's view, the sensor works at 10 Hz and it receives data at 10 Hz.

Reply:
@huhebo can you explain the use case? I can see the need to know what the actual ODR is, but I'm not seeing why we would want to know what the value we set it to.

Reply:
The `interval` is the data report interval for the sensor to its client, not the ODR of the sensor chip. 

The logic in `senss` is to hide internal arbitration to each application client, each client only need to set report interval value to the sensor via `senss_set_interval()` according to its requirements, and `senss_get_interval()` returns the report interval its set (each sensor have a `minimal report interval` attribute controls the fasted interval it can be set). 

This makes things clearer and simpler, from application client's view, it owns the sensor, no need care about if have other client(s) accessing the same sensor.   
   
Can you share the need to know what the actual ODR?